blocks/serialization.py

```diff
@@ -1,9 +1,119 @@
+
+    - The :func:`dump` strives to catch situations when the user tries
+      to pickle a function or a class not defined in the global namespace
+      and give a meaningful warning.
+
+If briefly, this module proposes a dumping mechanism which allows for
+greater robustness and persistency than standard pickling.
+
+Examples
+--------
+Consider a standard main loop (without an algorithm and a data stream
+for brevity)
+
+>>> from theano import tensor
+>>> from blocks.main_loop import MainLoop
+>>> from blocks.bricks import MLP, Tanh, Softmax
+>>> from blocks.model import Model
+>>> mlp = MLP([Tanh(), None], [784, 10, 10])
+>>> x = tensor.matrix('features')
+>>> y = tensor.lmatrix('targets')
+>>> cost = Softmax().categorical_cross_entropy(
+...            y.flatten(), mlp.apply(tensor.flatten(x, outdim=2)))
+>>> main_loop = MainLoop(None, None, model=Model(cost))
+
+Let's see how the main loop is dumped by :func:`dump`
+
+>>> from blocks.serialization import dump, load
+>>> import tarfile
+>>> with open('main_loop.tar', 'wb') as dst:
+...     dump(main_loop, dst)
+>>> tarball = tarfile.open('main_loop.tar', 'r')
+>>> tarball # doctest: +ELLIPSIS
+<tarfile.TarFile object at ...>
+>>> tarball.getnames()
+['_pkl']
+>>> tarball.close()
+
+As promised, the dump is a tarball. Since we did not ask for any additional
+magic, it just contains the pickled main loop in '_pkl' file.
+
+Let's do something more interesting:
+
+>>> with open('main_loop.tar', 'wb') as dst:
+...     dump(main_loop, dst,
+...          parameters=main_loop.model.parameters)
+>>> tarball = tarfile.open('main_loop.tar', 'r')
+>>> tarball.getnames()
+['_parameters', '_pkl']
+
+As requested by specifying the `_parameters` argument, the parameters were
+saved in a zip file.
+
+>>> import numpy
+>>> ps = numpy.load(tarball.extractfile(tarball.getmember('_parameters')))
+>>> sorted(ps.keys()) # doctest: +ELLIPSIS
+['|mlp|linear_0.W', '|mlp|linear_0.b', '|mlp|linear_1.W', '|mlp|lin...]
+>>> ps.close()
+
+The names for parameters are chosen intellegently to reflect their
+position in the brick hierarchy, if they belong to bricks, and by
+simply using the `.name` attribute, if they do not.
+
+The loading of the main loop as a whole still works:
+
+>>> with open('main_loop.tar', 'rb') as src:
+...     main_loop_loaded = load(src)
+>>> main_loop_loaded # doctest: +ELLIPSIS
+<blocks.main_loop.MainLoop object at ...>
+
+Additionally, this module provides convenience routine
+:func:`load_parameters`:
+
+>>> with open('main_loop.tar', 'rb') as src:
+...     parameters = load_parameters(src)
+>>> sorted(parameters.keys()) # doctest: +ELLIPSIS
+['/mlp/linear_0.W', '/mlp/linear_0.b', '/mlp/linear_1.W', '/mlp/line...]
+
+Loading parameters saved by :func:`dump` with :func:`load_parameters`
+ensures that their heirarchical names are compatible with
+:class:`~blocks.model.Model` and :class:`~blocks.select.Selector` classes.
+
+TODO: Add information about :func:`add_to_dump`.
```

Comment:
It would be great to also mention here `pickle_object` option, because it is a great new feature that this PR brings. This, together with documenting `add_to_dump`, can be done in the ticket though.